QA-0.01 온보딩 약관동의 화면 에러 및 회원탈퇴 Response, 로그인 화면 말풍선 등 수정

derrickkim0109 opened the pull request:
## PR 타입
어떤 변경에 대한 PR인가요?

<!-- 아래 체크리스트 중 해당 되는 부분에 체크해주세요 "x". -->

- [x] 버그 수정
- [ ] Feature / 신기능
- [ ] 코드 스타일 업데이트 (formatting, local variables)
- [ ] 리팩토링 (no functional changes, no api changes)
- [ ] 빌드 관련 변경사항
- [ ] CI 관련된 변경사항
- [ ] Documentation content changes
- [ ] 구조 변경
- [ ] Other... 추가 설명 요망:


## 배경

<!-- 해당 PR에서 개발한 내용에 대해 알아야 하는 배경지식을 작성해주세요. -->

**온보딩**

- 온보딩 Flow의 이용약관 동의 화면에서 전체 동의 선택 > 다른 동의 체크 박스 선택 시 `시작하기` 버튼이 해제되지 않는 버그
- 온보딩 프로필 설정에서 다음 버튼을 여러 번 선택해서 API 콜이 여러 번 진행되어 error Alert 발생
- `마지막에 로그인했어요` 말풍선 아래 뽀족한 부분 border 다크모드 대응되지 않았음
- 온보딩 프로필 설정 후, 회원 가입 Action 버튼 클릭 시 2초동안 API 콜 안되도록 throttle 처리
- 회원가입 Error 메세지 정의

**노트 작성 / 수정**

- 400자 이상 작성시 스크롤이 깨지는 문제
- 글자수와 본문 텍스트 겹치는 문제
- 노트 수정 화면 진입 시 가사 배경이 적용되지 않는 문제
- 가사 배경 텍스트가...으로 축약되는 문제
- 노트 수정에서 가사 배경, 본문만 수정해도 완료 버튼이 활성화 되게 해야함
- 곡 선택 후,  아티스트 정보 보이는 뷰 flex layout이랑 겹쳐서 레이아웃 깨지는 문제
- 수정 진입 시 유저가 선택한 가사배경 설정안되는 문제

**곡 검색**
- 리스트 중 레이블 텍스트가 긴 경우, 레이아웃 깨지는 문제
- viewWillAppear 시점에 맞춰 곡 검색 업데이트 되도록 수정

**마이페이지**
- 회원탈퇴 Response 프로퍼티 이름 변경됨
- 아티스트 셀 border 컬러가 변경되는 문제
- 알림 아이콘 위치 맞지 않는 문제
- 서비스 문의하기 링크 연결
- 프로필 수정에서 아이콘만 수정안되는 문제
- 회원 정보 수정(출생연도, 성별) 출생연도 혹은 성별 수정 시 버튼 활성화되도록 수정
- 회원 정보 수정(출생연도, 성별) 변경되지 않는 문제
- viewWillAppear 시점에 맞춰 내 노트, 북마크 업데이트 되도록 수정
- 서비스 배너 링크 추가
- 탭 진입 시 유저 아이콘 설정안되는 문제

**그외**
- 탭바 백그라운드 컬러 디자인과 다른 문제


## 목표

<!-- 해당 PR에서 개발/수정 한 기능의 목표를 작성해주세요. -->

- 각 화면에서 발생하는 버그들을 수정해야 함
- 가장 큰 부분은 노트 작성 / 수정에서 노트 텍스트 뷰가 키보드가 올라온 상태에서 계속 개행이 추가되어 내려갈 때, 
패딩 값에 맞게 맞춰서 스크롤이 내려가도록 구현되야 함

## 결과 (Optional)

<!-- 개발한 내용의 결과를 설명해주세요. !-->

| 노트 작성 | 노트 수정 - 키보드 개행|
|--|--|
|<img src="https://github.com/user-attachments/assets/9d12fe57-aca0-40c8-a402-1f65a1624cf0" width="250">|<img src="https://github.com/user-attachments/assets/7e3ed62a-6935-484b-b70a-98619e0f8f4c" width="250">|

| 다크모드에서 말풍선 색 통일 | 약관동의|
|--|--|
|<img src="https://github.com/user-attachments/assets/e650f429-e03e-4147-8809-e7272ccc3749" width="250">|<img src="https://github.com/user-attachments/assets/f69b55c7-d505-4f72-990c-50cdf2920b61" width="250">|

| 회원탈퇴|온보딩 프로필 설정에서 다음 버튼 Throttle 처리|
|--|--|
|<img src="https://github.com/user-attachments/assets/69f9de11-31e2-4398-a9d2-1789c8fbb651" width="250">|<img src="https://github.com/user-attachments/assets/724de1c2-6119-465b-ba82-706b39281682" width="250">|

# 공유

- 프로필 수정을 기존과 동일하게 작성해도 성공 Response를 내려줌 -> API 변경 필요
  - 동일 데이터에 대한 에러케이스 처리 필요함
- 출생연도, 성별 변경 API 에러 발생 수정 필요

derrickkim0109 committed:
refactor: DeleteUser API Response 객체 프로퍼티 변경

derrickkim0109 committed:
fix: 가입 동의 체크박스 전체 선택 후 한개만 해제 시 시작하기 버튼 해제되도록 수정

derrickkim0109 committed:
fix: 마지막에 로그인했어요. 다크모드 대응

derrickkim0109 committed:
fix: 온보딩 Flow 프로필 설정 후, 다음 버튼 클릭 시 2초동안 API콜 안되도록 수정

inwoodev commented on Projects/Core/Network/Interface/Sources/DTO/Response/FeelinDefaultResponse.swift:
확인했습니다~

inwoodev commented on Projects/Feature/Onboarding/Interface/Sources/ViewControllers/ProfileViewController.swift:
```diff
@@ -71,6 +71,7 @@ public final class ProfileViewController: UIViewController {
             .eraseToAnyPublisher()
 
         let nextButtonPublisher = nextButton.publisher(for: .touchUpInside)
+            .throttle(for: .seconds(2), scheduler: RunLoop.main, latest: false)
             .eraseToAnyPublisher()
```
태현님, 혹시 에러가 발생하면  coordinator를 통해서 화면전환이 되지 않는 부분도 테스트되셨을 까요?

그리고 @jcrescent61 이 첨부해 주신 gif를 확인 했을 때 에러 메세지 또한 정상적으로 출력되지 않는 듯 합니다.

그래서 제가 드리고 싶은 의견은
1. `SignUpError` 모델 수정(FeelinError 메세지를 출력할 수 있게. 자세한 건 다른 에러들 참고하시면 좋을 듯 합니다)
2. 에러 발생 시 화면 이동 방지
3. 그리고 throttle 대신에 `switchToLatest`가 이 상황에서는 더 요긴하게 중복 네트워크 콜 호출을 방지할 수 있지 않나 싶습니다!
p3

derrickkim0109 replied:
넵 알겠습니다~!

derrickkim0109 replied:
1. SignUpError의 message를 추가했습니다.
2. 이 부분은 이미 이동을 방지하고 있습니다.
3. 엘렌이 해주신 QA는 여러 번 시작하기 버튼을 클릭하여 이미 API를 호출한 상태에서 발생할 수 있는 문제로 이해했는데 제가 잘못 본 걸까요? 제 생각에는 2번이 가능했던 이유는 여러 번 API를 호출해서 발생한 문제라고 생각합니다. 

처음에 회원가입 API 호출을 한 상태에서 또 다시 SignUp API를 호출하여 두 번째 API 호출로 받은 Response는 Error를 방출하여 Alert 에러 메세지를 받은 상황으로 이해했습니다.
그래서 제 생각에는 switchToLatest 처리는 지금 상황과는 맞지않은 대처라고 생각이 드는데 어떻게 생각하시나요? 

음.. 제 생각에는 switchToLatest는 가장 마지막에 방출된 Publisher를 받게 하는 기능이라고 하면, 결국 서버에서는 회원가입 완료로 id가 발급되서 계속해서 Error를 보내게 되서 회원가입 완료로 넘어가지 못하게 될 것 같습니다.
그래서 throttle로 처음에 호출한 API를 2초동안 기다리도록 하는 것으로 처리해봤습니다. 

제가 이해한게 맞을까요?

derrickkim0109 replied:
success를 받는 경우, nextButton의 isEnable false로 처리하는 건 어떻게 생각하세요?

inwoodev replied:
요 일단 throttle 부분은 제가 잘못 이해한게 맞습니다 ㅎㅎ; switchToLatest나 throttle 둘 다 사용해도 무방한듯 합니다. 혼란을 야기해서 죄송합니다 :)

inwoodev replied:
`success를 받는 경우, nextButton의 isEnable false로 처리하는 건 어떻게 생각하세요?` 이 의견도 좋습니다~~

derrickkim0109 replied:
아닙니다~ 

구두로 논의한대로, 현 부분은 throttle로 유지하도록 하겠습니다.

inwoodev reviewed (changes requested):
고생하셨습니다~ 이전 pr커밋이 추가되지 않은듯 한데 요 부분 확인 해 주시면 좋을 듯 합니다~ 그리고 코멘트 남겼어요. 확인 해 주세요~

derrickkim0109 committed:
refactor: SignUpError 에러 메세지 정의

derrickkim0109 committed:
refactor: 특정 시점에 맞게 데이터 업데이트

- 곡검색
- 마이페이지 내노트, 북마크

derrickkim0109 committed:
remove: 노트 작성 Post API 콜 시 불필요한 mapError 처리 제거

derrickkim0109 committed:
style: 개행 추가

derrickkim0109 committed:
fix: 곡 검색 UI 레이아웃 넘어가는 부분 수정

derrickkim0109 committed:
refactor: addTrackLabel flex layout으로 겹치도록 변경

derrickkim0109 committed:
refactor: 노트 작성 뷰에 키보드 높이 설정 제거 및 각 뷰컨으로 이동

derrickkim0109 committed:
remove: 불필요한 설정 제거

derrickkim0109 committed:
fix: 노트 수정화면에서 유저가 선택한 가사 배경이 보이도록 수정

- 가사배경 버튼 클릭 시 선택했던 배경으로 셀 선택되도록 수정

derrickkim0109 committed:
add: 서비스 배너 링크 추가

derrickkim0109 committed:
fix: 마이페이지 문제 수정

- 탭 진입 시 유저 아이콘 설정되도록 수정
- 프로필 수정 > 유저 아이콘 만 변경 시 변경 가능하도록 수정
- 프로필 수정 후 마이페이지에서 변경된 아이콘 적용되도록 수정
- 프로필 수정 > 키보드 return 클릭 시 키보드 내려가도록 수정

derrickkim0109 committed:
remove: 불필요한 에러 처리 및 타입 변경

derrickkim0109 committed:
refactor: 노트 작성 / 수정 뷰컨 부분 곡 검색 버튼  + 버튼 라인 전체 선택가능하도록 수정

- 키보드 내림 처리 구현

derrickkim0109 committed:
fix: 마이페이지 알림 버튼 패딩 수정

derrickkim0109 committed:
fix: 마이페이지 아티스트 셀 border 컬러 변경되는 에러 수정

derrickkim0109 committed:
feat: 마이페이지 북마크 탭 아티스트 리스트 불러오는 API 추가

derrickkim0109 committed:
style: ProfileEditViewModel 메서드 프로퍼티 네이밍 변경

derrickkim0109 committed:
fix: 노트 수정 페이지 가사 배경 / 가사 작성 / 노트 작성에 완료 활성화 되도록 수정

- ProfileEditVieWModel 내 불필요한 코드 제거

derrickkim0109 committed:
fix: 노트 작성 / 수정 가사 작성 텍스트 뷰 분기처리

- 텍스트 없이 개행만 있을 경우 개행 2줄을 넘지 못함
- 50자를 넘으면 작성 못함
- 2줄 개행 후, 텍스트는 텍스트 뷰에 설정한 width값을 넘지 못함

derrickkim0109 committed:
refactor: 신고하기에서 텍스트 뷰가 활성화된 후 스크롤 시 키보드 내려가도록 수정

- 불필요한 설정 제거
- 개행 변경
- 노트 작성/수정 뷰 이미지 아이콘 radius 추가
- LyricsContentsView 레이아웃 설정 위치 수정

derrickkim0109 committed:
fix: 노트 작성 텍스트 뷰 각 분기별로 spacer View 동적으로 높이 조절

derrickkim0109 committed:
refactor: placeholder 넣지 않도록 수정

derrickkim0109 committed:
refactor: 에러 메세지로 변경

derrickkim0109 committed:
feat: 노트 작성 / 수정 후 노트 업데이트를 위한 메서드 추가

derrickkim0109 committed:
refactor: 출생, 성별 모델 DTO 수정

derrickkim0109 committed:
refactor: 출생연도 dropbutton 기본 값 추가

derrickkim0109 committed:
refactor: 불필요한 코드 제거 및 collectionView delegate로 수정

derrickkim0109 committed:
refactor: 출생연도, 성별 수정 API 콜 로직 수정

derrickkim0109 committed:
refactor: 가사 배경 선택 레이아웃 깨지는 부분 비율로 height 변경

derrickkim0109 committed:
fix: 아티스트 셀 선택 후 각 노트 데이터 받을 수 있도록 수정

derrickkim0109 committed:
fix: get Public Notification 내부 실행 메서드 오타 수정

derrickkim0109 committed:
fix: TabBar 컬러 변경

derrickkim0109 committed:
feat: safeArea Bottom 크기 구하는 프로퍼티 구현

derrickkim0109 committed:
refactor: 노트 작성 / 수정 뷰 노트 텍스트 뷰 각 분기별로 크기 조절 되도록 수정

derrickkim0109 committed:
refactor: 노트 작성 버튼 크기 56에 고정

derrickkim0109 committed:
add: 가사배경 텍스트 뷰 로직 추가

derrickkim0109 committed:
remove: 불필요한 코드 제거

derrickkim0109 committed:
refactor: 노트 작성 / 수정에서 선택된 아티스트 곡 있을 시 표기되는 아이콘 수정

derrickkim0109 committed:
refactor: 노트 텍스트 뷰 다이나믹하게 높이 조절되도록 수정

derrickkim0109 committed:
refactor: 노트 작성 버튼 크기 조절

derrickkim0109 committed:
fix: 비회원 시 안보이던 텍스트 수정

derrickkim0109 commented on Projects/Feature/Home/Interface/Sources/Views/WritingNote/WritingNoteView.swift:
```diff
@@ -351,4 +351,25 @@ public final class WritingNoteView: UIView {
         artistNameLabel.flex.markDirty()
     }
+
+    public func calculateNoteCharCountContainerHeight(
```
기존에 말씀드렸던 방식은 note 텍스트뷰의 라인이 변경되는 만큼 screen Height의 비율로 spacer를 설정했었습니다. 
하지만 이 방식은 처리할 수 없는 분기처리가 너무 많다는 문제가 있었습니다.

그래서 스크린 높이에 noteContainerView가 차지해야 할 부분을 제외한 나머지 부분을 빼는 방식으로 구현해봤습니다.

derrickkim0109 commented on Projects/Coordinator/Home/Interface/Sources/HomeCoordinatorInterface.swift:
```diff
@@ -162,6 +162,10 @@ extension HomeCoordinator: CoordinatorDelegate,
                            SearchSongViewControllerDelegate {
+    public func didFinishForPresentingViewController() {
+        dismissViewController()
+        // TODO: 노트 작성 완료 후 업데이트 메서드 호출
```
Postnote, Editnote 완료 이후 시점입니다.

derrickkim0109 commented on Projects/Feature/Home/Interface/Sources/ViewControllers/WritingNote/PostNoteViewController.swift:
```diff
@@ -211,16 +236,43 @@ public final class PostNoteViewController: UIViewController {
+                if keyboardHeight > 0 {
+                    let noSafeArea = safeAreaBottomInset == 0
+                    // 33 = safeAreaBottom + 12(카운트 bottom)
+                    let additionalBottomInset: CGFloat = noSafeArea ? 33 : 0
```
se의 경우 safe Area bottom을 가져올 수 없어서 하드코딩했습니다.

derrickkim0109 replied:
12는 note container 의 bottom 값입니다.

derrickkim0109 commented on Projects/Feature/Home/Interface/Sources/ViewControllers/WritingNote/PostNoteViewController.swift:
```diff
@@ -305,18 +355,23 @@ public final class PostNoteViewController: UIViewController {
+                let lines = text.components(separatedBy: .newlines)
+                let numberOfLines = lines.count
+
+                if (text.count == 0 || text.count < Const.lyricsMaxTextLength) && numberOfLines > maxLineCount {
+                    let truncatedText = lines.dropLast().joined(separator: "\n")
+                    lyricsTextView.text = truncatedText
+                } else if text.count > Const.lyricsMaxTextLength {
+                    lyricsTextView.text = String(text.prefix(Const.lyricsMaxTextLength))
+                } else if lyricsTextView.isThirdLineExceedingWidth() {
+                    lyricsTextView.text = String(text.dropLast(2))
+                } else {
+                    setupLyricsTextviewTextCenterVertically(lyricsTextView)
+                    updateCharacterCountForLyrics()
```
기존에 delegate `func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool`에 선언한 부분을 didChange로 로직을 옮겼습니다.

inwoodev replied:
현 상황에서는 이게 최선인것 같네요. 고생하셨습니다~

derrickkim0109 commented on Projects/Feature/Home/Interface/Sources/Views/WritingNote/WritingNoteView.swift:
```diff
@@ -218,45 +220,33 @@ public final class WritingNoteView: UIView {
-        contentView.addSubview(addTrackLabel)
-        addSubview(noteCharCountLabel)
-
-        NSLayoutConstraint.activate([
-            addTrackLabel.centerYAnchor.constraint(
-                equalTo: iconImageView.centerYAnchor
-            ),
-            addTrackLabel.leadingAnchor.constraint(
-                equalTo: iconImageView.trailingAnchor,
-                constant: 10
-            )
-        ])
```
기존에 구현할 때는 겹치도록 구현하는 방법을 생각하지 못하고 개발하여 Constraint으로 위에 뷰를 겹치는 방식을 선택했지만, 이 설정으로 flexlayout이 깨지는 문제가 있어서 flexlayout으로 변경했습니다.

derrickkim0109 commented on Projects/Feature/Home/Interface/Sources/Views/WritingNote/WritingNoteView.swift:
```diff
@@ -326,65 +335,14 @@ public final class WritingNoteView: UIView {
         artistNameLabel.text = item.artist.name
 
-        addToPlayButton.setImage(FeelinImages.play, for: .normal)
+        addToPlayButton.setImage(FeelinImages.selectedSongActive, for: .normal)
```
기획, 디자인 변경으로 아이콘을 변경했습니다.

inwoodev commented on Projects/Feature/Home/Interface/Sources/Views/WritingNote/WritingNoteView.swift:
```diff
@@ -351,4 +351,25 @@ public final class WritingNoteView: UIView {
+        screenHeight: CGFloat,
+        keyboardHeight: CGFloat
+    ) -> CGFloat {
+        let navigationHeight: CGFloat = 44
+        let bottom: CGFloat = 12
+        let safeAreaBottom: CGFloat = 21
+
+        let contentHeight = screenHeight - pin.safeArea.top - navigationHeight - safeAreaBottom - bottom
+        let rootScrollViewMargin: CGFloat = 16.0
+        let artistInfoHeaderViewHeight: CGFloat = 40 + 24 + 2 + 16
+        let lyricsTextBodyViewHeight: CGFloat = 132 + 20 + 12 + 32
+        let noteTextViewContainerHeight: CGFloat = 24 - noteTextView.frame.height
+        let minimumHeight: CGFloat = 24
+
+        // 모든 값을 제외하고 55가 모잘라서 넣음
+        let remainingHeight: CGFloat = contentHeight - rootScrollViewMargin - artistInfoHeaderViewHeight - lyricsTextBodyViewHeight - noteTextViewContainerHeight - minimumHeight - keyboardHeight + 55
+
+        return max(remainingHeight, minimumHeight)
+    }
```
오 괜찮은 접근인듯 합니다. 다만 추후 artistInfoHeaderViewHeight와 lyricsTextBodyViewHeight를 하드코딩하지 않는 방향으로 수정 해 볼 수 있을 듯 해요~

p5

inwoodev reviewed (approved):
야근하시느라 바쁘셨을텐데...와 너무 고생하셨습니다.

머지해도 괜찮을 듯 하네요.
